File: Sorting Algorithms/bubble_sort.cpp
/*
Given an array, the elements of the array are to be arranged in an increasing order.

Algorithm:
1. Traverse the array, compare each element with the adjacent element,
2. swap them if(cur_element > next_element) else continue the traversal
3. Repeat the above step till the array is completely sorted.

Input:
5 1 4 2 3

Output:
1 2 3 4 5

*/

#include<iostream>

using namespace std;

void bubbleSort(int arr[], int size)
{
    int temp;
    for(int i=0;i<size-1;i++)
    {
        for(int j=0;j<size-i-1;j++)
        {
            if(arr[j]>arr[j+1])
            {
                temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
            }
        }
    }
}

int main(int argc, char const *argv[])
{
    int arr[] = {5, 1, 4, 2, 3};

    int size;

    size = sizeof(arr)/sizeof(arr[0]);
    bubbleSort(arr,size);

    for(int ind=0;ind<size;ind++)
        cout<<arr[ind]<<" ";
    return 0;
}

File: Sorting Algorithms/Bubble_Sort.java
/*
Bubble Sort is the simplest sorting algorithm
that works by repeatedly swapping the adjacent
elements if they are in wrong order.


*/
public class BubbleSortExample {  
    static void bubbleSort(int[] arr) {  
        // finding array size. 
        int n = arr.length;  
        int temp = 0;  
        //each iteration place max element of 1 to n-i index at index n-i
         for(int i=0; i < n; i++){  
                 for(int j=1; j < (n-i); j++){  
                          if(arr[j-1] > arr[j]){  
                                 //swap elements  
                                 temp = arr[j-1];  
                                 arr[j-1] = arr[j];  
                                 arr[j] = temp;  
                         }   
                 }  
         }  
  
    }  
    public static void main(String[] args) {  
                int arr[] ={3,60,35,2,45,320,5};  
                 
                System.out.println("Array Before Bubble Sort");  
                for(int i=0; i < arr.length; i++){  
                        System.out.print(arr[i] + " ");  
                }  
                System.out.println();  
                  
                bubbleSort(arr);//sorting array elements using bubble sort  
                 
                System.out.println("Array After Bubble Sort");  
                for(int i=0; i < arr.length; i++){  
                        System.out.print(arr[i] + " ");  
                }  
   
        }  
}  

/* 
Bubble Sort always runs O(n^2) time even if the array is sorted. 
It can be optimized by stopping the algorithm if inner loop didn’t cause any swap. 
*/
